File: mariadb-api/src/Services/ProjectSettingsSerializer.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

/**
 * Ported from Services/ProjectSettingsSerializer.cs. Defaults mirror normalizeHeaderButtonVisibility
 * (src/js/storage.js) exactly: every field is opt-out (defaults true) except workflow,
 * changeAuditing and retrospective, which are opt-in (default false) — a missing/corrupted value
 * must never silently start enforcing/recording/showing something the user never asked for. Keys
 * are camelCase to match both the frontend's own field names and the "changeAuditing" key
 * TaskService::isChangeAuditingEnabled reads from this same column.
 *
 * @phpstan-type ProjectSettings array{documents:bool,risks:bool,decisions:bool,health:bool,principles:bool,objectives:bool,teamsCommittees:bool,workflow:bool,timeTracking:bool,changeAuditing:bool,subTasks:bool,retrospective:bool,strategy:bool,dashboards:bool,forms:bool,portfolioPlanner:bool,portals:bool,resources:bool}
 */
final class ProjectSettingsSerializer
{
    private const DEFAULTS = [
        'documents' => true,
        'risks' => true,
        'decisions' => true,
        'health' => true,
        'principles' => true,
        'objectives' => true,
        'teamsCommittees' => true,
        'workflow' => false,
        'timeTracking' => true,
        'changeAuditing' => false,
        'subTasks' => true,
        // Opt-in, like workflow: brand-new functionality nobody has configured yet, so a
        // missing/corrupted value must never silently turn it on.
        'retrospective' => false,
        // Opt-in, like workflow/retrospective: a missing/corrupted value must never silently turn on
        // a module the project never asked for.
        'strategy' => false,
        // Opt-in, like strategy: the Self-Service Dashboard module stays invisible until a Project
        // Admin deliberately turns it on.
        'dashboards' => false,
        // Opt-in, Org-Admin-authored: Enterprise Forms & Workflow stays invisible until an Org Admin
        // deliberately turns it on for this project (same shape as strategy).
        'forms' => false,
        // Opt-in, same shape as forms — was previously a pure Org-Admin permission gate with no
        // per-project toggle at all; a missing/corrupted value must fail closed to that same
        // hidden-until-toggled behavior, not silently re-expose the nav entry everywhere.
        'portfolioPlanner' => false,
        // Opt-in, same shape as forms/portfolioPlanner — Organisational Portals stays invisible
        // until an Org Admin deliberately turns it on for this project.
        'portals' => false,
        // Opt-in, same shape as forms/portfolioPlanner/portals — the org-wide Resources
        // utilisation-over-time chart stays invisible until an Org Admin deliberately turns it on.
        'resources' => false,
    ];

    public static function serialize(array $settings): string
    {
        $result = [];
        foreach (self::DEFAULTS as $key => $default) {
            $result[$key] = (bool) ($settings[$key] ?? $default);
        }
        return json_encode($result);
    }

    /** @return ProjectSettings */
    public static function parse(?string $json): array
    {
        $decoded = [];
        if ($json !== null && $json !== '') {
            $tmp = json_decode($json, true);
            if (is_array($tmp)) {
                $decoded = $tmp;
            }
        }

        $result = [];
        foreach (self::DEFAULTS as $key => $default) {
            $value = $decoded[$key] ?? null;
            $result[$key] = is_bool($value) ? $value : $default;
        }
        return $result;
    }
}

File: api/Enkl.Api/Services/ProjectSettingsSerializer.cs
using System.Text.Json;
using Enkl.Api.Dtos;

namespace Enkl.Api.Services;

/// <summary>
/// Shared camelCase (de)serialization for Project.HeaderButtonVisibilityJson. Defaults mirror
/// normalizeHeaderButtonVisibility (src/js/storage.js) exactly: every field is opt-out (defaults to
/// true, so a missing/corrupted value never silently hides something the user never chose to hide)
/// except Workflow, ChangeAuditing and Retrospective, which are opt-in (default false, so a
/// missing/corrupted value never silently starts enforcing/recording/showing something the user
/// never asked for).
/// </summary>
public static class ProjectSettingsSerializer
{
    private static readonly JsonSerializerOptions Options = new(JsonSerializerDefaults.Web);

    public static string Serialize(ProjectSettingsDto settings) => JsonSerializer.Serialize(settings, Options);

    public static ProjectSettingsDto Parse(string? json)
    {
        JsonDocument? doc = null;
        try
        {
            if (!string.IsNullOrWhiteSpace(json)) doc = JsonDocument.Parse(json);
        }
        catch (JsonException)
        {
            // Corrupted/garbled JSON falls through to defaults below, same as the client-side guard.
        }

        bool Get(string name, bool defaultValue)
        {
            if (doc is not null && doc.RootElement.TryGetProperty(name, out var val) &&
                (val.ValueKind == JsonValueKind.True || val.ValueKind == JsonValueKind.False))
            {
                return val.ValueKind == JsonValueKind.True;
            }
            return defaultValue;
        }

        var result = new ProjectSettingsDto(
            Documents: Get("documents", true),
            Risks: Get("risks", true),
            Decisions: Get("decisions", true),
            Health: Get("health", true),
            Principles: Get("principles", true),
            Objectives: Get("objectives", true),
            TeamsCommittees: Get("teamsCommittees", true),
            Workflow: Get("workflow", false),
            TimeTracking: Get("timeTracking", true),
            ChangeAuditing: Get("changeAuditing", false),
            SubTasks: Get("subTasks", true),
            // Opt-in, like Workflow: brand-new functionality nobody has configured yet, so a
            // missing/corrupted value must never silently turn it on.
            Retrospective: Get("retrospective", false),
            // Opt-in, like Workflow/Retrospective: a missing/corrupted value must never silently
            // turn on a module the project never asked for.
            Strategy: Get("strategy", false),
            // Opt-in, like Strategy: the Self-Service Dashboard module (nav entry, picker, viewer/
            // editor) stays invisible until a Project Admin deliberately turns it on.
            Dashboards: Get("dashboards", false),
            // Opt-in, Org-Admin-authored: Enterprise Forms & Workflow stays invisible until an Org
            // Admin deliberately turns it on for this project (same shape as Strategy).
            Forms: Get("forms", false),
            // Opt-in, same shape as Forms — was previously a pure Org-Admin permission gate with no
            // per-project toggle at all; a missing/corrupted value must fail closed to that same
            // hidden-until-toggled behavior, not silently re-expose the nav entry everywhere.
            PortfolioPlanner: Get("portfolioPlanner", false),
            // Opt-in, same shape as Forms/PortfolioPlanner — Organisational Portals stays invisible
            // until an Org Admin deliberately turns it on for this project.
            Portals: Get("portals", false),
            // Opt-in, same shape as Forms/PortfolioPlanner/Portals — the org-wide Resources
            // utilisation-over-time chart stays invisible until an Org Admin deliberately turns it on.
            Resources: Get("resources", false));

        doc?.Dispose();
        return result;
    }
}
